### PR TITLE
Add DashboardLayout component to @toolpad/core (old, PR has been split)

### PR DESCRIPTION
**Pull request by apedroferreira**
Docs/demo: https://deploy-preview-3343--mui-toolpad-docs.netlify.app/toolpad/core/components/dashboard-layout/

(once we have a navigation adapter we can make the demo sidebar links work with local state to provide better examples)
(API docs generation to be added next in separate PR)

---

Ready with component tests, supports `branding` with logo and name for header, and `navigation` with nested items in sidebar. Over time we can expand on this and add more features / improvements, of course.

Please check the types in [packages/toolpad-core/src/AppProvider/AppProvider.tsx](https://github.com/mui/mui-toolpad/pull/3343/files#diff-d72d0f71caf7ef5c3f7a2250be04d58d3dc741a4e944a9c4a9f5bf1f518e6b5f) + component props ([packages/toolpad-core/src/AppProvider/AppProvider.tsx](https://github.com/mui/mui-toolpad/pull/3343/files#diff-d72d0f71caf7ef5c3f7a2250be04d58d3dc741a4e944a9c4a9f5bf1f518e6b5f) and [packages/toolpad-core/src/layout/DashboardLayout/DashboardLayout.tsx](https://github.com/mui/mui-toolpad/pull/3343/files#diff-16914c8076bb1bbac0a0309370f80d3a528d427b0e7dd9c62c4472575f4760d7)) for the chosen API.

For now, `AppProvider` can be imported as a named import from `@toolpad/core/AppProvider` and `DashboardLayout` the same way from `@toolpad/core/DashboardLayout`. They can also be imported directly from `@toolpad/core` along with some maybe useful types such as `Branding` and `Navigation`.

Made it so that `pnpm dev` can be run to build Toolpad core + watch changes, and then a new Next.js project in `playground/toolpad-core-nextjs` can be run to develop and see changes live. We can add other projects/frameworks to the playground folder later.

Added documentation page for the DashboardLayout component with demos. The demos can also be used for development but in some cases a real application is better, of course.

Also added simple contribution instructions for Toolpad Core.

### CODE REVIEW

File: packages/toolpad-core/src/layout/DashboardLayout.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+const DRAWER_WIDTH = 240;
+
+interface DashboardLayoutProps {
+  children: React.ReactNode;
+}
+
+export default function DashboardLayout({ children }: DashboardLayoutProps) {
+  return (
+    <NoSsr>
```
Comment (apedroferreira) [inline packages/toolpad-core/src/layout/DashboardLayout.tsx:36]: Theming isn't working without `NoSsr`, trying to understand these types of issues better there seem to be a lot of them...

File: packages/toolpad-core/src/client/AppProvider/AppProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  navigation,
+}: AppProviderProps) {
+  return (
+    <AppRouterCacheProvider options={{ enableCssLayer: true }}>
+      <ThemeProvider theme={theme}>
+        {/* CssBaseline kickstarts an elegant, consistent, and simple baseline to build upon. */}
+        <CssBaseline />
+        <BrandingContext.Provider value={branding}>
+          <NavigationContext.Provider value={navigation}>
+            {/* @TODO: Can probably remove NoSsr once we have non CSS-in-JS solution. */}
```
Comment (Janpot) [inline packages/toolpad-core/src/client/AppProvider/AppProvider.tsx:31]: MUI should be able to ssr https://mui.com/material-ui/guides/server-rendering/

File: packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  List,
+  ListItem,
+  ListItemButton,
+  ListItemIcon,
+  ListItemText,
+  ListSubheader,
+  Stack,
+  Toolbar,
+  Typography,
+} from '@mui/material';
```
Comment (oliviertassinari) [inline packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx:18]: flat imports

File: packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+// Must use named imports or app crashes when the component is imported in a Next.js app
+// eslint-disable-next-line no-restricted-imports
+import { ExpandLess as ExpandLessIcon, ExpandMore as ExpandMoreIcon } from '@mui/icons-material';
```
Comment (oliviertassinari) [inline packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx:21]: Flast imports

File: packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
+/**
+ * @vitest-environment jsdom
+ */
+
+import * as React from 'react';
+import { describe, test, expect, afterEach } from 'vitest';
```
Comment (oliviertassinari) [inline packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.spec.tsx:6]: Test files should be `.test.tsx`. `spec.tsx` is for TypeScript tests.

File: packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+import { styled } from '@mui/system';
+import Link from 'next/link';
```
Comment (oliviertassinari) [inline packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx:23]: No coupling with any router

```suggestion
```

Comment (Janpot) [reply packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx:23]: @apedroferreira perhaps to get a baseline version merged we can start with plain `a` tags? Then progressively improve with a pluggable router adapter and/or Link component?

Comment (Janpot) [reply packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx:23]: What I've been prototyping with is an [adapter](https://github.com/Janpot/dashboard-framework-prototype/blob/412060d835deffc3bb348e2f77bbf7ae1291d945/src/lib/dash/filter.tsx#L134-L188) that is modeled as a subset of the new [navigation API](https://developer.mozilla.org/en-US/docs/Web/API/Navigation_API). That way hopefully we can some day just delegate directly to `window.navigation` as the default router adapter. It's a challenge though to bridge to this from the different Next.js routers.

Comment (apedroferreira) [reply packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx:23]: I'm also using `@mui/material-nextjs/v14-appRouter` but the pages router should use `'@mui/material-nextjs/v14-pagesRouter'` and Vite shouldn't need anything... So I guess we need to allow the user to choose a framework somehow and have a component that uses one of these according to that setting.

Comment (Janpot) [reply packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx]: that is what the top level provider is for

Comment (apedroferreira) [reply packages/toolpad-core/src/client/layout/DashboardLayout/DashboardLayout.tsx]: I made this version framework agnostic for now but if I use `<a>` tags I need to use something like `window.location.pathname` to compare routes, and this shows SSR errors in Next.js.
I can work on some kind of framework-aware navigation adapter in this PR already to solve this issue, or we can merge this and I can work on it next.

**Comment by apedroferreira** [comment]
I've added review fixes and some discussed changes such as a different build process using babel and a playground for development.
I've also updated the PR description accordingly.

Not really sure what is happening with the errors in CI, let me know if you do!

File: packages/create-toolpad-app/src/generateProject.ts [hunk 7]
```diff
@@ -139,24 +100,66 @@ export default function generateProject(
...
+
+  const inter = Inter({
     weight: ["300", "400", "500", "700"],
     subsets: ["latin"],
     display: "swap",
   });
-  
-  const theme = createTheme({
+
+  const baseTheme = createTheme({
```
Comment (apedroferreira) [inline packages/create-toolpad-app/src/generateProject.ts:113]: It would be ideal to have a design to match this theme to, I kind of made it look like the other template we have but not always sure about the right colors for example.

File: packages/toolpad-core/babel.config.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
+const path = require('path');
```
Comment (apedroferreira) [inline packages/toolpad-core/babel.config.js:1]: Got these mostly from Base UI, let me know if there's anything important we should be doing differently as I don't fully understand these build scripts.

File: scripts/build.mjs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+          description: `Valid bundles: "${validBundles.join('" | "')}"`,
+          type: 'string',
+        })
+        .option('largeFiles', {
+          type: 'boolean',
+          default: false,
+          describe: 'Set to `true` if you know you are transpiling large files.',
+        })
+        .option('out-dir', { default: './build', type: 'string' })
+        .option('watch', { type: 'boolean' })
```
Comment (apedroferreira) [inline scripts/build.mjs:120]: I've added this `watch` argument so we can build on changes during development.

**Comment by Janpot** [comment]
I think to prove that the build is working correctly, this PR should show that it can be displayed in a demo component in the docs as well as in  the Toolpad studio application.

File: packages/toolpad-core/src/types/Navigation.types.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  icon: React.ReactNode;
+  routes?: (Omit<NavigationRoute, 'routes'> & { path: string })[];
+}
+
+export interface NavigationSection {
+  title: string;
+  routes: NavigationRoute[];
+}
+
+export type Navigation = NavigationSection[];
```
Comment (Janpot) [inline packages/toolpad-core/src/types/Navigation.types.ts:13]: When I designed the interface to represent navigation I tried to optimize for a variety of use case:

flat list of pages

```
- page 1
- page 2
- page 3
```
=>
```tsx
[
  { name: 'page-1', title: 'Page 1' },
  { name: 'page-2', title: 'Page 2' },
  { name: 'page-3', title: 'Page 3' }
]
```
flat list with sub headers

```
header 1
------
- page 1
- page 2
header 2
------
- page 3
```
=>
```tsx
[
  { kind: 'header', title: 'Header 1' },
  { name: 'page-1', title: 'Page 1' },
  { name: 'page-2', title: 'Page 2' },
  { kind: 'header', title: 'Header 2' },
  { name: 'page-3', title: 'Page 3' }
]
```
deeply nested

```
- page 1
- folder 1
  - page 2
  - folder 2
    - folder 3
      - page 3
- folder 3
  - page 4
```
=>
```tsx
[
  { name: 'page-1', title: 'Page 1' },
  { 
    name: 'folder-1', 
    title: 'Folder 1',
    children: [
      { name: 'page-2', title: 'Page 2' },
      {
        name: 'folder-2',
        title: 'Folder 2',
        children: [
          { 
            name: 'folder-3', 
            title: 'Folder 3',
            children: [
              { name: 'page-3', title: 'Page 3' },
            ]
          }
        ]
      }
    ]
  },
  { 
    name: 'folder-3', 
    title: 'Folder 3',
    children: [
      { name: 'page-4', title: 'Page 4' }
    ]
  }
]
```

and any combination of pages, folders, and subheaders.

I believe the interface I came up with is the simplest structure that supports all these use cases. Can you give examples for each of those with the interfaces you've designed?

File: packages/toolpad-core/src/types/Navigation.types.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+export interface NavigationRoute {
```
Comment (Janpot) [inline packages/toolpad-core/src/types/Navigation.types.ts:1]: I think I prefer to keep the word "route" out of the nomenclature, this is not a router

File: packages/toolpad-core/src/types/index.ts [hunk 12]
```diff
@@ -0,0 +1,2 @@
+export type * from './Branding.types';
```
Comment (Janpot) [inline packages/toolpad-core/src/types/index.ts:1]: Why the .types.ts extension? And is it really necessary to create a separate top level folder for types? I feel like in core/X types are grouped logically with their components.

I feel like it would make most sense to start like MUI core, where all components are exported top level. And we can start organizing them once we're reaching a sizable amount. wdyt?

File: packages/toolpad-core/src/context/BrandingContext.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+import * as React from 'react';
```
Comment (Janpot) [inline packages/toolpad-core/src/context/BrandingContext.ts:1]: I think we should try to avoid grouping things by function (e.g. "contexts", "modules", "types", "controllers", "views"). That doesn't scale.
If it's internal only, I'd nest it under the `AppProvider`. If we will export this as an API, I think we can do like base does it, under a folder named as the hook that will eventually consume this context: https://github.com/mui/base-ui/tree/master/packages/mui-base/src/useTabs

File: packages/toolpad-core/src/types/Navigation.types.ts (hunk 10 above)
Comment (apedroferreira) [reply packages/toolpad-core/src/types/Navigation.types.ts:13]: The interface I came up with supports multiple lists with different titles too, and up to one level of nesting. 
It would be easy to make it infinitely nestable, but a second level of nesting already starts looking a bit cramped in the sidebar.

Example with 2 lists and an item with nested items (I think this covers all the cases?):
```
const NAVIGATION: Navigation = [
  {
    title: 'Primary items',
    routes: [
      {
        label: 'Dashboard1',
        path: '/dashboard1',
        icon: <DashboardIcon />,
      },
      {
        label: 'Orders1',
        path: '/dashboard/orders1',
        icon: <ShoppingCartIcon />,
        items: [
          {
            label: 'Nested Orders 1',
            path: '/dashboard/orders1/nested1',
            icon: <ShoppingCartIcon />,
          },
          {
            label: 'Nested Orders 2',
            path: '/dashboard/orders1/nested2',
            icon: <ShoppingCartIcon />,
          },
        ]
      },
    ],
  },
  {
    title: 'Secondary items',
    items: [
      {
        label: 'Dashboard 2',
        path: '/dashboard2',
        icon: <DashboardIcon />,
      },
      {
        label: 'Orders2',
        path: '/dashboard/orders2',
        icon: <ShoppingCartIcon />,
      },
    ],
  },
];
```

I can make the titles optional so that not every list needs a title.
I guess it covers the same cases as your suggestion? I find my format easier to read but i don't mind changing it.
One con of your suggestion would maybe be that the relation between a header and its items is not so clear (it's determined by the order in the array).

Comment (Janpot) [reply packages/toolpad-core/src/types/Navigation.types.ts]: > I think this covers all the cases?

How does this format support a flat list of pages without sections or subfolders?

> One con of your suggestion would maybe be that the relation between a header and its items is not so clear (it's determined by the order in the array).

The relation is similar to the relation between titles in html and paragraphs. It's not a hierarchical relationship, the headers are to be seen as markup. The flat list keeps it extensible for adding more markup in the future

Comment (apedroferreira) [reply packages/toolpad-core/src/types/Navigation.types.ts]: > How does this format support a flat list of pages without sections or subfolders?

If we don't want a title I could titles optional and it would be:

```
const NAVIGATION: Navigation = [
  {
    items: [
      {
        label: 'Dashboard1',
        path: '/dashboard1',
        icon: <DashboardIcon />,
      },
      {
        label: 'Orders1',
        path: '/dashboard/orders1',
        icon: <ShoppingCartIcon />,
      },
    ],
  },
];
```

So yeah there's the `items` property that feels a bit extra in this case, could be seen as a con.

> The relation is similar to the relation between titles in html and paragraphs. It's not a hierarchical relationship, the headers are to be seen as markup. The flat list keeps it extensible for adding more markup in the future. (e.g. look at the MUI docs, one could imagine the sidebar ads are another type of markup)

True, it's more extensible, didn't consider that. I don't mind going with it, shouldn't even be a big change!

Comment (Janpot) [reply packages/toolpad-core/src/types/Navigation.types.ts]: > If we don't want a title I could make title optional so it would be:

I'm not a fan of having the "title" have both the function of overwriting the name, as well as turn on/off certain features.
How do you differentiate between a folder and a subheader then? One being collapsible and the other separating groups of pages on the same level?


🤔 A horizontal separator is probably a better example for another markup element than the ad

Comment (apedroferreira) [reply packages/toolpad-core/src/types/Navigation.types.ts]: The title does not turn on/off features, it's just the heading for each top-level list of navigation items and shows up above them.
Collapsible items are navigation items with the `items` property themselves.
The dividers show up automatically between navigation lists.
I added a screenshot to the PR where this shows.

In the header navigation I assume we could use the same system but no titles.

Anyway let's go with your approach.

Comment (Janpot) [reply packages/toolpad-core/src/types/Navigation.types.ts]: > Collapsible items are navigation items with the items property themselves.

I'm still not fully getting it. How do you differentiate between a section and a folder? Aren't both of them navigation items with the `items` property?

e.g. subheader nested in a folder, or vice-versa:

```
- page1
subheader1
- page2
-- divider --
- page3
- folder1
  - page4
  - subheader2
  - folder2
    - page5
    -- divider --
    - page6
  - page7
```

> The dividers show up automatically between navigation lists.

Ok, but that doesn't allow for dividers between any two arbitrary items.

Comment (apedroferreira) [reply packages/toolpad-core/src/types/Navigation.types.ts]: > I'm still not fully getting it. How do you differentiate between a section and a folder? Aren't both of them navigation items with the items property?

In my first approach it's only possible to add headings at the top-level (they're just the `title`), so not inside collapsible items, so you can't have sections inside folders...
So I guess that your approach in more versatile in that way. I'm trying it out!

Comment (Janpot) [reply packages/toolpad-core/src/types/Navigation.types.ts]: Oh I see, your top-level is different from nested levels

Comment (apedroferreira) [reply packages/toolpad-core/src/types/Navigation.types.ts]: I've also just realized that using your suggested approach for subheaders can be tricky with the MUI List component https://mui.com/material-ui/react-list/ since it takes a `subheader` as a prop for the list.
We can probably use subheaders without the prop, we'll just lose the automatic sticky headers.

Comment (Janpot) [reply packages/toolpad-core/src/types/Navigation.types.ts]: > since it takes a subheader as a prop for the list.

The content of this prop has to be wrapped in a `<ListSubHeader>` component, and these can also be used inside of the list, just like `<Divider>`. See https://mui.com/material-ui/react-list/#sticky-subheader.

Not 100% sure what would be the best approach. We could declare a navigation item with a title and children but no declared path segment as a section. OTOH a flat list could be more versatile towards the future.

File: docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md [hunk 14]
```diff
@@ -22,9 +22,9 @@ The `navigation` prop in the `AppProvider` allows for setting any type of naviga
 
 {{"demo": "DashboardLayoutNavigation.js", "height": 640, "iframe": true}}
 
-## API
+<!-- ## API
```
Comment (apedroferreira) [inline docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:25]: Let's leave API docs generation to another PR as it doesn't seem super simple and we can discuss how to do it.

**Comment by apedroferreira** [comment]
Not sure about the remaining CI errors - looks like we may need bigger instances for Circle CI if it's running out of memory?
And not sure why the Netlify build fails, `pnpm docs:build` works locally, not sure if it's some cache that can be cleared or something.

File: .circleci/config.yml [hunk 15]
```diff
@@ -62,7 +62,7 @@ commands:
             pnpm --version
       - run:
           name: Install js dependencies
-          command: pnpm install
+          command: pnpm install --ignore-scripts
```
Comment (apedroferreira) [inline .circleci/config.yml:65]: This seems to fix the out of memory issues in CI but might be making some integrations tests fail, looking more into it.

Comment (Janpot) [reply .circleci/config.yml:65]: I tried a few updates to this branch, cleaned up the dependencies a bit and removed the playground folder. That branch seems to pass CI https://github.com/apedroferreira/mui-toolpad/pull/1

Comment (apedroferreira) [reply .circleci/config.yml:65]: I see, yeah I guess some of the babel packages I copied over might not be needed...
As for the playground, if we get rid of it do you have any alternative in mind? When I tried to make Toolpad Core work in development in projects outside the monorepo I always had some weird issues.

Comment (Janpot) [reply .circleci/config.yml]: We can try figuring out which part of the playground package.json is the culprit by bissecting the dependencies and  see what happens in CI. Timeconsuming, but probably the only way

Comment (apedroferreira) [reply .circleci/config.yml]: It was probably Next.js having a different version than the docs, should be good now.

**Comment by apedroferreira** [comment]
Submitting for review again, CI is finally fixed.

File: .npmrc [hunk 16]
```diff
@@ -1,5 +1,5 @@
 # Enforce pnpm
 
-engine-strict = true
-auto-install-peers = true
+engine-strict=true
+auto-install-peers=true
 strict-peer-dependencies=false
```
Comment (oliviertassinari) [inline .npmrc:5]: It seems to me that a space is supposed to be present from https://docs.npmjs.com/cli/v9/configuring-npm/npmrc.

File: .eslintignore [hunk 17]
```diff
@@ -4,6 +4,7 @@ pnpm-lock.yaml
 /docs/export
 /docs/schemas
 /packages/**/dist
+/packages/**/build
```
Comment (oliviertassinari) [inline .eslintignore:7]: Could be a global rule?

```suggestion
build
```

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBranding.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    icon: <ShoppingCartIcon />,
+  },
+];
+
+const BRANDING = {
+  logo: <img src="https://mui.com/static/logo.png" alt="MUI logo" />,
+  title: 'MUI',
+};
+
+export default function DashboardLayoutBranding() {
```
Comment (oliviertassinari) [inline docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBranding.js:26]: Missing TypeScript version?

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+import { BrandingContext, Navigation, NavigationContext, NavigationPageItem } from '../AppProvider';
+
+const IS_CLIENT = typeof window !== 'undefined';
+
+const DEFAULT_DRAWER_WIDTH = 320;
+
+const LogoContainer = styled('div')({
+  position: 'relative',
+  height: 40,
+  img: {
```
Comment (oliviertassinari) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:30]: Does this mean?

```suggestion
  '& img': {
```

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+import Box from '@mui/material/Box';
+import Typography from '@mui/material/Typography';
+import DashboardIcon from '@mui/icons-material/Dashboard';
+import ShoppingCartIcon from '@mui/icons-material/ShoppingCart';
+import BarChartIcon from '@mui/icons-material/BarChart';
+import DescriptionIcon from '@mui/icons-material/Description';
+import LayersIcon from '@mui/icons-material/Layers';
+import { AppProvider } from '@toolpad/core/AppProvider';
+import { DashboardLayout } from '@toolpad/core/DashboardLayout';
+import theme from '../theme';
```
Comment (oliviertassinari) [inline docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js:11]: It don't think we should have relative imports for demos, it will break once exported to codesandbox.

**Comment by oliviertassinari** [comment]
It feels like some of the changes in this PR could be extracted to another one that isn't about the DashboardLayout component so that other people can work in parallel on other building blocks.

File: packages/create-toolpad-app/src/generateProject.ts [hunk 21]
```diff
@@ -9,84 +9,44 @@ export default function generateProject(
...
+  import type { Navigation } from "@toolpad/core";
   import theme from '../theme';
+
+  const NAVIGATION: Navigation = [
+    {
+      kind: 'header',
+      title: 'Main items',
+    },
+    {
+      path: '/page',
```
Comment (Janpot) [inline packages/create-toolpad-app/src/generateProject.ts:23]: I believe a good guiding principle in designing this interface would be to make sure that the individual items are easy to move around, or copy. Just the same way how folders on a file system are easy to move around. The disign principle makes sense since very often these navigation structures will mimic a folder structure in the project.

By making these `path` parameters absolute, moving them around will be more difficult. i.e. moving a nested item under one top-level to an other will require you to rewrite all paths of nested items underneath.

Instead I'd considder letting navigation items only define the segment for which they are responsible, and build the full path by chaining them recursively.

There may be use-cases where users may want to target a completely different path than the one in the hierarchy, but this should be the special case. I wouldn't tackle this until a user makes a compelling case on why this is strictly necessary.

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+}: DashboardSidebarSubNavigationProps) {
+  // Rerender once to update with client-side navigation
+  const [hasMounted, setHasMounted] = React.useState(false);
+
+  const [expandedSidebarItemIds, setExpandedSidebarItemIds] = React.useState<string[]>([]);
+
+  React.useEffect(() => {
+    setHasMounted(true);
+
+    if (IS_CLIENT) {
```
Comment (Janpot) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:52]: Why is this needed? Effects don't run serverside.

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+        return (
+          <React.Fragment key={navigationItemId}>
+            {navigationItem.path ? (
+              <a href={navigationItem.path} style={{ color: 'inherit', textDecoration: 'none' }}>
+                {listItem}
+              </a>
+            ) : (
+              listItem
+            )}
+            {IS_CLIENT && hasMounted && navigationItem.children ? (
```
Comment (Janpot) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:142]: `IS_CLIENT` can never be `false` if `hasMounted` is `true`.


```suggestion
            {hasMounted && navigationItem.children ? (
```

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+  subNavigation: Navigation;
+  depth?: number;
+}
+
+function DashboardSidebarSubNavigation({
+  subNavigation,
+  depth = 0,
+}: DashboardSidebarSubNavigationProps) {
+  // Rerender once to update with client-side navigation
+  const [hasMounted, setHasMounted] = React.useState(false);
```
Comment (Janpot) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:45]: I'm not a big fan of this method. This will prevent rendering certain elements on the first render, even if it's never being hydrated. Have you tried the `useIsSsr` hook that's available in `@mui/utils`?

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+        const nestedNavigationCollapseIcon = isNestedNavigationExpanded ? (
+          <ExpandLessIcon />
+        ) : (
+          <ExpandMoreIcon />
+        );
+
+        const listItem = (
+          <ListItem>
+            <ListItemButton
+              selected={IS_CLIENT && hasMounted && window.location.pathname === navigationItem.path}
```
Comment (Janpot) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:123]: Same as before.

```suggestion
              selected={hasMounted && window.location.pathname === navigationItem.path}
```

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 26]
```diff
@@ -0,0 +1,233 @@
...
+          </React.Fragment>
+        );
+      })}
+    </List>
+  );
+}
+
+interface DashboardLayoutProps {
+  children: React.ReactNode;
+  sidebarWidth?: number;
```
Comment (Janpot) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:159]: Does this really have to be configurable? And if it were configurable, wouldn't it make sense to be responsive? What if we just come up with good defaults based on the breakpoint and wait for making it configurable on user feedback?

Does the drawer collapse on narrower screens?

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx [hunk 27]
```diff
@@ -0,0 +1,233 @@
...
+  hideTitle?: boolean;
+}
+
+export function DashboardLayout({
+  children,
+  sidebarWidth = DEFAULT_DRAWER_WIDTH,
+  hideTitle = false,
```
Comment (Janpot) [inline packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:166]: I feel like at this point it's hard to say whether this behavior should be enabled through a specific prop, or we should let it be solved by slots API.
Perhaps at this stage we should concentrate more on coming up with the best possible default experience, for all breakpoints and think of customization once we've collect a few use cases.

File: packages/toolpad-core/src/AppProvider/AppProvider.tsx [hunk 28]
```diff
@@ -1,23 +1,52 @@
...
+
+export type Navigation = NavigationItem[];
+
+export const BrandingContext = React.createContext<Branding | null>(null);
+
+export const NavigationContext = React.createContext<Navigation>([]);
+
 interface AppProviderProps {
   children: React.ReactNode;
   theme: Theme;
```
Comment (Janpot) [inline packages/toolpad-core/src/AppProvider/AppProvider.tsx:38]: Does it have to be mandatory? We could provide a default.
```suggestion
  theme?: Theme;
```

File: packages/toolpad-core/src/AppProvider/AppProvider.tsx [hunk 29]
```diff
@@ -1,23 +1,52 @@
...
+  branding?: Branding | null;
+  navigation: Navigation;
```
Comment (Janpot) [inline packages/toolpad-core/src/AppProvider/AppProvider.tsx:40]: Default navigation can be `[]`? i.e. I should be allowed to drop in a dashboard layout and concentrate on the refining the branding before adding pages. Or vice versa.

```suggestion
  navigation?: Navigation;
```

File: packages/toolpad-core/.npmrc [hunk 30]
```diff
@@ -0,0 +1 @@
+enable-pre-post-scripts=true
```
Comment (Janpot) [inline packages/toolpad-core/.npmrc:1]: Do we need this? I see no pre or post scripts in the package.json. And we probably shouldn't rely on them anyway if the package manager doesn't support it out of the box. We'll do task orchestration with `nx` if we need to. X/core will be using it vor that purpose as well in the future.

File: packages/toolpad-core/babel.config.js [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+
+const productionPlugins = [
+  ['babel-plugin-react-remove-properties', { properties: ['data-mui-test'] }],
+];
+
+module.exports = function getBabelConfig(api) {
+  const useESModules = api.env(['regressions', 'stable', 'rollup']);
+
+  const defaultAlias = {
+    '@toolpad/core': resolveAliasPath('./packages/toolpad-core/src'),
```
Comment (Janpot) [inline packages/toolpad-core/babel.config.js:19]: No aliasing in babel. it's not the right tool for that. I see we're using `vitest` for testing, if you need aliasing, do it in `vitest`.

You can also remove the error codes macro. We'll add this optimization later if we need it. We're going to figure out a more portable method to support it.

In fact `babel.config.js` is only for top-level configuration. If possible, we should remove this file altogether.

File: packages/toolpad-core/package.json [hunk 32]
```diff
@@ -1,75 +1,90 @@
...
   },
+  "sideEffects": false,
   "publishConfig": {
     "access": "public"
   },
   "engines": {
     "node": ">=18"
-  }
+  },
+  "files": [
```
Comment (Janpot) [inline packages/toolpad-core/package.json:84]: Aren't we building into a `build` folder? Like core?

File: package.json [hunk 33]
```diff
@@ -14,7 +14,7 @@
     "markdownlint": "markdownlint-cli2 \"**/*.md\"",
     "prettier": "pretty-quick --ignore-path .eslintignore",
     "prettier:all": "prettier --write . --ignore-path .eslintignore",
-    "dev": "dotenv cross-env FORCE_COLOR=1 lerna -- run dev --stream --parallel --ignore docs",
+    "dev": "dotenv cross-env FORCE_COLOR=1 lerna -- run dev --stream --parallel --ignore docs --ignore toolpad-core-nextjs",
```
[the remark this answers is not in the document]
Comment (Janpot) [reply package.json:17]: We will move towards `nx` for orchestration over time

**Comment by Janpot** [comment]
> It feels like some of the changes in this PR could be extracted to another one that isn't about the DashboardLayout component so that other people can work in parallel on other building blocks. 

Agreed. The goal in the PR should be to set up the bare minimum of the basic layout component and provider so that we can test/prove the project setup works. I'd strip the configuration options from these components and concentrate on nailing the project setup so we can get this trough.
We can then concentrate on API subsequently.

**Comment by apedroferreira** [comment]
> > It feels like some of the changes in this PR could be extracted to another one that isn't about the DashboardLayout component so that other people can work in parallel on other building blocks.
> 
> Agreed. We could make the goal of the PR to set up the bare minimum of the basic layout component and provider so that we can test/prove the project setup works. Perhaps we should strip the configuration options from these components and concentrate on nailing the project setup so we can get this trough? We can then concentrate on API in follow up PR.

The current state just evolved naturally from review feedback + improving the build to fix some bugs with the component itself. I can probably split project setup from the component at this point though, at least, will see what else I can split.

File: packages/toolpad-core/package.json (hunk 32 above)
Comment (apedroferreira) [reply packages/toolpad-core/package.json:84]: The base repo uses a `build` folder but this is how the files are listed there.

File: docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js (hunk 20 above)
Comment (apedroferreira) [reply docs/data/toolpad/core/components/dashboard-layout/DashboardLayoutBasic.js:11]: Hm ok just wanted to reuse the same theme in all demos, will see how else it can be done.

File: packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx (hunk 26 above)
Comment (apedroferreira) [reply packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:159]: It was just a nice easy to add option, didn't look into responsiveness though.

> Does the drawer collapse on narrower screens?

No.

File: packages/toolpad-core/src/AppProvider/AppProvider.tsx (hunk 28 above)
Comment (apedroferreira) [reply packages/toolpad-core/src/AppProvider/AppProvider.tsx:38]: If you use `create-toolpad-app` it comes with a default theme, but it's not built into the library.
We could add it as a default theme though and not have it set by `create-toolpad-app`, i guess that sounds good.

Comment (Janpot) [reply packages/toolpad-core/src/AppProvider/AppProvider.tsx:38]: We can have it created by `create-toolpad-app`, but that doesn't mean this property has to be mandatory?

Comment (apedroferreira) [reply packages/toolpad-core/src/AppProvider/AppProvider.tsx:38]: First options I can think of:

1. Users can provide separate themes for light and dark mode;
2. Theme can take a function with `mode: PaletteMode` as an argument.

I was trying to keep it simple for now thinking we can still make breaking changes afterwards.
I guess option 2 should be more flexible as in option 1 it would be harder to change parts of the theme conditionally based on the mode. But probably we should also allow just passing in a theme in a straightforward way.

File: packages/toolpad-core/.npmrc (hunk 30 above)
Comment (apedroferreira) [reply packages/toolpad-core/.npmrc:1]: Changing buils script to ""build": "pnpm prebuild && pnpm build:node && pnpm build:stable && pnpm build:types && pnpm build:copy-files" if that's better.

File: packages/create-toolpad-app/src/generateProject.ts (hunk 21 above)
Comment (apedroferreira) [reply packages/create-toolpad-app/src/generateProject.ts:23]: I've done this but before it was easy to define if a collapsible navigation item links to a page or not itself by having a `path` property or not.
If we use the `path` of a collapsible item as the base path for the nested items inside it, we need to have the `path` there whether the collapsible item is a link or not, to serve as the base path for the nested items.

So I'm trying to find the best solution for this... Maybe even collapsible items should never be links themselves, not sure, or we could add a new property just for this but it may be overkill?

**Comment by apedroferreira** [comment]
Completed/split in other PRs.
